### ISSUE TITLE
Investigate potential Fasta reader performance regression

### ISSUE THREAD

**Issue by lbergelson** [samtools/htsjdk#471]
There was a report about terrible performance reading zipped fasta files in GATK.  Since no changes were made to FASTA reading in GATK itself it's possible that this is due to changes in htsjdk.  

Those with access can view it here: https://github.com/broadinstitute/dsde-docs/issues/570

The relevant report is:

```
I noticed a performance bottleneck with newer GATK versions (I'm on the latest version right now v3.5-0-g36282e4, Compiled 2015/11/25 04:03:56) where (simple) analysis types such as SelectVariants or VariantsToTable would run for days when they would previously complete in less than an hour. I have to add that I am using bgzip-compressed and tabix-indexed VCF files throughout the analysis and that I see the performance problem on our cluster running several processes in parallel. Most Java processes would show a rather low percentage of CPU consumption, while our cluster file system maxes out at around 100,000 read requests per second. Using 'strace' of an active Java thread, I noticed that millions of "pread" system calls of 183 bytes against the reference sequence *.fasta file seem to be issued.

  [ ~]$ strace -p 28395
  Process 28395 attached - interrupt to quit
  pread(11, "TGGTAAAATGTGGTTGGATGAAGCGTACGCTT"..., 183, 46038724) = 183
  pread(11, "TAGACCATTCTATCAAAATGCTCTTTCTACAG"..., 183, 46038907) = 183
  ...
The question is, what am I doing wrong? Alternatively, are you aware of any changes to how the reference sequence FASTA file is read in recent GATK versions especially in the context of bgzip-compressed and tabix-indexed VCF files. Are you aware that so many read requests are issued? I tried to go back to version v3.3-0-g37228af, Compiled 2014/10/24 01:07:22 and although the "pread" requests seem to include larger regions, the run time of a single process seems similar. I would have to go back further, but you may be quicker in doing so.

Suggested by our system administrator, I managed to overcome the performance bottleneck by copying the reference sequence *.fasta, *.fasta.fai and *.dict files into the Linux shared memory tmpfs file system (/dev/shm/). As an example, VariantsToTable run times have fallen from 5 hours to just 31 seconds (13h vs 74s per 1M sites), SelectVariants from 5 hours to 18 minutes (26m vs 87s per 1M sites). In case I'm not the only one seeing this, would it make sense to read (and keep in memory) larger chunks of the reference sequence?
```

We did recently touch the fasta reader in  4f550e1f1afabf21467957fa672ca2a4ad457897, we should make sure we haven't caused a performance regression.

**Comment by droazen** [comment]
Will discuss this with @cmnbroad today and see if he has time to look into this.

**Comment by droazen** [comment]
The original report is: http://gatkforums.broadinstitute.org/gatk/discussion/6977/large-number-of-pread-system-calls-on-the-reference-sequence-fasta-file/p1

**Comment by cmnbroad** [comment]
It looks like the current release of GATK is still on htsjdk 1.141, and the nio Path change was first released in 2.0.0, so I don't think its that. It does look like there have been a couple of other changes that could affect this though. Next step is to verify that.

**Comment by cmnbroad** [comment]
It looks like https://github.com/samtools/htsjdk/commit/e6ba234b47050bc037a1023daed64350d6e11209 caused a performance regression. It greatly reduced the size of the read buffer, thereby increasing the number of system calls required to retrieve a large sequence. This is especially dramatic in the case of the GATK CachingIndexedFastaSequenceFile, which retrieving and caching megabase sized shards. Before this change it would take about 8 calls to cache 1M bases; with the change it looks like it takes hundreds of calls.

**Comment by droazen** [comment]
Ouch -- in other words, that change was designed to improve performance when retrieving small sequences, and in the process it killed performance for retrieving large sequences? What do you suggest we do here?

**Comment by tfenne** [comment]
@cmnbroad & @droazen I think that's just a bug.  The offending line:
```java
ByteBuffer channelBuffer = ByteBuffer.allocate(Math.min(Defaults.NON_ZERO_BUFFER_SIZE, (length % basesPerLine + 2) * bytesPerLine));
```
should be dividing, not mod'ing right?
```
ByteBuffer channelBuffer = ByteBuffer.allocate(Math.min(Defaults.NON_ZERO_BUFFER_SIZE, (length / basesPerLine + 2) * bytesPerLine));
```

**Comment by cmnbroad** [comment]
@tfenne Exactly right.

**Comment by droazen** [comment]
@tfenne Good catch -- mod makes no sense at all there.

**Comment by cmnbroad** [comment]
FWIW, if I traced the history right, this bug has been in htsjdk since 1.133 and in GATK since 3.2.

**Comment by lbergelson** [comment]
Should we consider adding performance tests?  It would be nice to have an automated way of verifying we haven't made any horrible performance regressions.

**Comment by cmnbroad** [comment]
The fix for what we believe to be the regression that caused this went in with https://github.com/samtools/htsjdk/pull/479.
